File: src/core/Account/Events.cs
using System;
using core.Shared;

namespace core.Account
{
    public class UserConfirmed : AggregateEvent
    {
        public UserConfirmed(Guid id, Guid aggregateId, DateTimeOffset when) : base(id, aggregateId, when)
        {
        }
    }

    public class UserCreated : AggregateEvent
    {
        public UserCreated(Guid id, Guid aggregateId, DateTimeOffset when, string email, string firstname, string lastname)
            : base(id, aggregateId, when)
        {
            Email = email;
            Firstname = firstname;
            Lastname = lastname;
        }

        public string Email { get; }
        public string Firstname { get; }
        public string Lastname { get; }
    }

    public class UserDeleted : AggregateEvent
    {
        public UserDeleted(Guid id, Guid aggregateId, DateTimeOffset when, string feedback) : base(id, aggregateId, when)
        {
            Feedback = feedback;
        }

        public string Feedback { get; }
    }

    [Obsolete("dropped this functionality from domain")]
    internal class UserLoggedIn : AggregateEvent
    {
        public UserLoggedIn(Guid id, Guid aggregateId, DateTimeOffset when) : base(id, aggregateId, when)
        {
        }
    }

    public class UserPasswordResetRequested : AggregateEvent
    {
        public UserPasswordResetRequested(Guid id, Guid aggregateId, DateTimeOffset when) : base(id, aggregateId, when)
        {
        }
    }

    public class UserPasswordSet : AggregateEvent
    {
        public UserPasswordSet(Guid id, Guid aggregateId, DateTimeOffset when, string hash, string salt)
            : base(id, aggregateId, when)
        {
            Hash = hash;
            Salt = salt;
        }

        public string Hash { get; }
        public string Salt { get; }
    }

    public class UserSubscribedToPlan : AggregateEvent
    {
        public UserSubscribedToPlan(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            string planId,
            string customerId,
            string subscriptionId)
            : base(id, aggregateId, when)
        {
            PlanId = planId;
            CustomerId = customerId;
            SubscriptionId = subscriptionId;
        }

        public string PlanId { get; }
        public string CustomerId { get; }
        public string SubscriptionId { get; }
    }

    public class UserConnectedToBrokerage : AggregateEvent
    {
        public UserConnectedToBrokerage(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            string accessToken,
            string refreshToken,
            string tokenType,
            long expiresInSeconds,
            string scope,
            long refreshTokenExpiresInSeconds)
            : base(id, aggregateId, when)
        {
            AccessToken = accessToken;
            RefreshToken = refreshToken;
            TokenType = tokenType;
            ExpiresInSeconds = expiresInSeconds;
            Scope = scope;
            RefreshTokenExpiresInSeconds = refreshTokenExpiresInSeconds;
        }

        public string AccessToken { get; }
        public string RefreshToken { get; }
        public string TokenType { get; }
        public long ExpiresInSeconds { get; }
        public string Scope { get; }
        public long RefreshTokenExpiresInSeconds { get; }
    }

    public class UserRefreshedBrokerageConnection : AggregateEvent
    {
        public UserRefreshedBrokerageConnection(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            string accessToken,
            string refreshToken,
            string tokenType,
            long expiresInSeconds,
            string scope,
            long refreshTokenExpiresInSeconds)
            : base(id, aggregateId, when)
        {
            AccessToken = accessToken;
            RefreshToken = refreshToken;
            TokenType = tokenType;
            ExpiresInSeconds = expiresInSeconds;
            Scope = scope;
            RefreshTokenExpiresInSeconds = refreshTokenExpiresInSeconds;
        }

        public string AccessToken { get; }
        public string RefreshToken { get; }
        public string TokenType { get; }
        public long ExpiresInSeconds { get; }
        public string Scope { get; }
        public long RefreshTokenExpiresInSeconds { get; }
    }

    public class UserDisconnectedFromBrokerage : AggregateEvent
    {
        public UserDisconnectedFromBrokerage(Guid id, Guid aggregateId, DateTimeOffset when) : base(id, aggregateId, when)
        {
        }
    }

    public class UserSettingSet : AggregateEvent
    {
        public UserSettingSet(Guid id, Guid aggregateId, DateTimeOffset when, string key, string value) : base(id, aggregateId, when)
        {
            Key = key;
            Value = value;
        }

        public string Value { get; }
        public string Key { get; }
    }
}

File: src/core2.fs/Account/Events.fs
namespace core.Account

open System
open core.Shared

[<AllowNullLiteral>]
type UserConfirmed(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)

[<AllowNullLiteral>]
type UserCreated(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, email: string, firstname: string, lastname: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Email = email with get
    member val Firstname = firstname with get
    member val Lastname = lastname with get

[<AllowNullLiteral>]
type UserDeleted(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, feedback: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Feedback = feedback with get

[<Obsolete("dropped this functionality from domain")>]
[<AllowNullLiteral>]
type internal UserLoggedIn(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)

[<AllowNullLiteral>]
type UserPasswordResetRequested(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)

[<AllowNullLiteral>]
type UserPasswordSet(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, hash: string, salt: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Hash = hash with get
    member val Salt = salt with get

[<AllowNullLiteral>]
type UserSubscribedToPlan(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, planId: string, customerId: string, subscriptionId: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val PlanId = planId with get
    member val CustomerId = customerId with get
    member val SubscriptionId = subscriptionId with get

[<AllowNullLiteral>]
type UserConnectedToBrokerage(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, accessToken: string, refreshToken: string, tokenType: string, expiresInSeconds: int64, scope: string, refreshTokenExpiresInSeconds: int64) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val AccessToken = accessToken with get
    member val RefreshToken = refreshToken with get
    member val TokenType = tokenType with get
    member val ExpiresInSeconds = expiresInSeconds with get
    member val Scope = scope with get
    [<Obsolete("this will not have a valid value and is here to preserve the event structure")>]
    member val RefreshTokenExpiresInSeconds = refreshTokenExpiresInSeconds with get

[<AllowNullLiteral>]
type UserRefreshedBrokerageConnection(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, accessToken: string, refreshToken: string, tokenType: string, expiresInSeconds: int64, scope: string, refreshTokenExpiresInSeconds: int64) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val AccessToken = accessToken with get
    member val RefreshToken = refreshToken with get
    member val TokenType = tokenType with get
    member val ExpiresInSeconds = expiresInSeconds with get
    member val Scope = scope with get
    member val RefreshTokenExpiresInSeconds = refreshTokenExpiresInSeconds with get

[<AllowNullLiteral>]
type UserDisconnectedFromBrokerage(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)

[<AllowNullLiteral>]
type UserSettingSet(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, key: string, value: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Key = key with get
    member val Value = value with get

[<AllowNullLiteral>]
type UserBrokerageInterestApplied(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, activityId: string, netAmount: decimal) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val ActivityId = activityId with get
    member val NetAmount = netAmount with get

[<AllowNullLiteral>]
type UserCashTransferApplied(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, activityId: string, netAmount: decimal) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val ActivityId = activityId with get
    member val NetAmount = netAmount with get
